gcc/cmassoc/html.1
.TH html 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
html - print HTML template page

.SH SYNOPSIS
.BR html
.RI [ options ]
.IR file 
.RI [ file ]
.RI [ ... ] 
.RI [> stdout ]

.PP
.BR html
.RI [ options ]
.RI [< stdin ]
.RI [> stdout ]

.SH DESCRIPTION
Read one or more text files and write them to stdout.
Prefix the text ststream with an HTML file header and suffix it with an HTML file footer.
This effectively insertsw the text stream into an HTML page, ready to display.

.SH OPTIONS
Long options are not supported.
All options are single letters.

.TP
.RB - o
Print a template profile on stdout.
The template uses default program values.
Output can be directed to a file, edited and read back when the program is next executed.
This option takes precedence over normal formatting operations.

.TP
-\fBp\fI profile\fR
The profile name.
If this option is present then the program attempts to open and read the profile.
The program will report an error and terminated if the named profile file does not exist.
If this option is omitted the program attempts to read the default profile.
The program does not report an error when the default profile is missing.
The default profile is "\fB/etc/html.ini\fR".

.TP
-\fBs\fI section\fR
The profile section name.
If this option is present the the program will search the named profile or default profile for the named section.
If this option is omitted the progam will search for the default section, instead.
The default section is named "\fBdefault\fR".

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over previous command line options.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over previous command line options.

.SH ARGUMENTS

.TP
\fIfile\fR
The name of a text file.
This file is copied, unmodified, to the stdout.

.SH PROFILE
The default profile looks like:

.PP
   [default]
   subject=Untitled Document
   contact=cmaier@cmassoc.net
   cssfile=/styles/cmassoc.css
   charset=iso-8859-1

.TP
.BR subject
The page title and level 1 heading.

.TP
.BR contact
The person or organization preparing the page.

.TP
.BR cssfile
The absolute or relative path to the CSS2 stylesheet.

.TP
.BR charset
The character set used to display the page.
This specification is appended to the content type specification in the XML process instruction and/or the HTML meta tag.

.SH EXAMPLES
The following example reads file \fBabc.txt\fR and writes it to stdout.
The file contains three numbered lines in this example.
File content is preceded with an HTML header and followed by an HTML footer.
Ouput appears on the console because we have not directed stdout to a file.

.PP
   # html abc.txt
   <html>
           <head>
                   <title>
                           untitled
                           </title>
                   <meta http-equiv='content-type' content='text/html;iso-8859-1'/>
                   <meta name='generator' content='Motley Tools 1.8.0'/>
                   <meta name='author' content='cmaier@cmassoc.net'/>
                   <meta name='robots' content='noindex,nofollow'/>
                   <link rel='stylesheet' href='/styles/cmassoc.css' type='text/css'/>
                   <style type='text/css'>
                           </style>
                   </head>
           <body class='standard'>
                   <div class='headerlink'>
                           [<a href='index.html' title='Index'>PREV</a>]
                           [<a href='index.html' title='Index'>HOME</a>]
                           [<a href='index.html' title='Index'>NEXT</a>]
                           </div>
                   <h1>
                           untitled
                           </h1>
                   <!-- BEGIN CONTENT -->
   1
   2
   3
                   <!-- END CONTENT -->
                   <div class='footerlink'>
                           [<a href='index.html' title='Index'>PREV</a>]
                           [<a href='index.html' title='Index'>HOME</a>]
                           [<a href='index.html' title='Index'>NEXT</a>]
                           </div>
                   </body>
           </html>

.SH SEE ALSO
.BR matrix ( 1 )
.BR offset ( 1 )

.SH CREDITS
 Charles Maier <cmaier@cmassoc.net>



gcc/cmassoc/html.c
/*====================================================================*
 *
 *   html.c - print HTML page template;
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/version.h"
#include "../tools/profile.h"
#include "../tools/markup.h"
#include "../tools/format.h"
#include "../tools/error.h"
#include "../tools/types.h"
#include "../tools/sizes.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/profile.c"
#include "../tools/totruth.c"
#include "../tools/indent.c"
#include "../tools/error.c"
#include "../tools/efreopen.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define HTML_PROFILE "/etc/html.ini"
#define HTML_SECTION "default"

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static char const * subject = HTML_SUBJECT; 
static char const * contact = HTML_AUTHOR; 
static char const * cssfile = CSS_STYLESHEET; 
static char const * charset = ISO_CHARSET; 

/*====================================================================*
 *
 *   unsigned links (unsigned margin, char const * style);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned links (unsigned margin, char const * style) 

{ 
	indent (margin++, "<div class='%s'>", style); 
	indent (margin, "[<a href='%s' title='%s'>%s</a>]", HTML_PATH, HTML_NAME, HTML_PREV); 
	indent (margin, "[<a href='%s' title='%s'>%s</a>]", HTML_PATH, HTML_NAME, HTML_HOME); 
	indent (margin, "[<a href='%s' title='%s'>%s</a>]", HTML_PATH, HTML_NAME, HTML_NEXT); 
	indent (margin--, "</div>"); 
	return (margin); 
} 

/*====================================================================*
 *
 *   unsigned header (unsigned margin);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned header (unsigned margin) 

{ 
	extern char const * contact; 
	extern char const * subject; 
	extern char const * cssfile; 
	extern char const * charset; 
	indent (margin++, "<html>"); 
	indent (margin++, "<head>"); 
	indent (margin++, "<title>"); 
	indent (margin, "%s", subject); 
	indent (margin--, "</title>"); 
	indent (margin, "<meta http-equiv='content-type' content='%s;%s'/>", ISO_CONTENT, charset); 
	indent (margin, "<meta name='author' content='%s'/>", contact); 
	indent (margin, "<meta name='generator' content='%s'/>", HTML_PROGRAM); 
	indent (margin, "<meta name='robots' content='%s'/>", HTML_ROBOTS); 
	indent (margin, "<link rel='stylesheet' href='%s' type='%s'/>", cssfile, CSS_CONTENT); 
	indent (margin++, "<style type='text/css'>"); 
	indent (margin--, "</style>"); 
	indent (margin--, "</head>"); 
	indent (margin++, "<body class='%s'>", "standard"); 
	margin = links (margin, STYLE_TOPLINK); 
	indent (margin++, "<h1>"); 
	indent (margin, "%s", subject); 
	indent (margin--, "</h1>"); 
	indent (margin, "<!-- BEGIN CONTENT -->"); 
	return (margin); 
} 

/*====================================================================*
 *
 *   unsigned body (unsigned margin);
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned body (unsigned margin) 

{ 
	signed c; 
	while ((c = getc (stdin)) != EOF) 
	{ 
		putc (c, stdout); 
	} 
	return (margin); 
} 

/*====================================================================*
 *
 *   unsigned footer (unsigned margin);
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned footer (unsigned margin) 

{ 
	indent (margin, "<!-- END CONTENT -->"); 
	margin = links (margin, STYLE_BOTLINK); 
	indent (margin--, "</body>"); 
	indent (margin--, "</html>"); 
	return (margin); 
} 

/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	extern char const * subject; 
	extern char const * contact; 
	extern char const * cssfile; 
	static char const * optv [] = 
	{ 
		"op:s:", 
		PUTOPTV_S_FUNNEL, 
		"print HTML page template", 
		"o\tprint default profile on stdout", 
		"p f\tuse profile (s) [" HTML_PROFILE "]", 
		"s s\tuse profile section (s) [" HTML_SECTION "]", 
		(char const *)(0)
	}; 
	char const * profile = HTML_PROFILE; 
	char const * section = HTML_SECTION; 
	unsigned margin = 0; 
	signed c; 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'o': 
			printf ("[%s]\n", section); 
			printf ("subject=%s\n", subject); 
			printf ("contact=%s\n", contact); 
			printf ("cssfile=%s\n", cssfile); 
			printf ("charset=%s\n", charset); 
			return (0); 
		case 'p': 
			profile = optarg; 
			break; 
		case 's': 
			section = optarg; 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	subject = profilestring (profile, section, "subject", subject); 
	contact = profilestring (profile, section, "contact", contact); 
	cssfile = profilestring (profile, section, "cssfile", cssfile); 
	charset = profilestring (profile, section, "charset", charset); 
	margin = header (margin); 
	if (!argc) 
	{ 
		margin = body (margin); 
	} 
	while ((argc) && (* argv)) 
	{ 
		if (efreopen (* argv, "rb", stdin)) 
		{ 
			margin = body (margin); 
		} 
		argc--; 
		argv++; 
	} 
	margin = footer (margin); 
	return (0); 
} 

